src/psbt.h

```diff
@@ -0,0 +1,562 @@
+    CTxOut witness_utxo;
+    CScript redeem_script;
+    CScript witness_script;
+    CScript final_script_sig;
+    CScriptWitness final_script_witness;
+    std::map<CPubKey, KeyOriginInfo> hd_keypaths;
+    std::map<CKeyID, SigPair> partial_sigs;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+    int sighash_type = 0;
+
+    bool IsNull() const;
+    void FillSignatureData(SignatureData& sigdata) const;
+    void FromSignatureData(const SignatureData& sigdata);
+    void Merge(const PSBTInput& input);
+    bool IsSane() const;
+    PSBTInput() {}
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+        // Write the utxo
+        // If there is a non-witness utxo, then don't add the witness one.
+        if (non_witness_utxo) {
+            SerializeToVector(s, PSBT_IN_NON_WITNESS_UTXO);
+            OverrideStream<Stream> os(&s, s.GetType(), s.GetVersion() | SERIALIZE_TRANSACTION_NO_WITNESS);
+            SerializeToVector(os, non_witness_utxo);
+        } else if (!witness_utxo.IsNull()) {
+            SerializeToVector(s, PSBT_IN_WITNESS_UTXO);
+            SerializeToVector(s, witness_utxo);
+        }
+
+        if (final_script_sig.empty() && final_script_witness.IsNull()) {
+            // Write any partial signatures
+            for (auto sig_pair : partial_sigs) {
+                SerializeToVector(s, PSBT_IN_PARTIAL_SIG, MakeSpan(sig_pair.second.first));
+                s << sig_pair.second.second;
+            }
+
+            // Write the sighash type
+            if (sighash_type > 0) {
+                SerializeToVector(s, PSBT_IN_SIGHASH);
+                SerializeToVector(s, sighash_type);
+            }
+
+            // Write the redeem script
+            if (!redeem_script.empty()) {
+                SerializeToVector(s, PSBT_IN_REDEEMSCRIPT);
+                s << redeem_script;
+            }
+
+            // Write the witness script
+            if (!witness_script.empty()) {
+                SerializeToVector(s, PSBT_IN_WITNESSSCRIPT);
+                s << witness_script;
+            }
+
+            // Write any hd keypaths
+            SerializeHDKeypaths(s, hd_keypaths, PSBT_IN_BIP32_DERIVATION);
+        }
+
+        // Write script sig
+        if (!final_script_sig.empty()) {
+            SerializeToVector(s, PSBT_IN_SCRIPTSIG);
+            s << final_script_sig;
+        }
+        // write script witness
+        if (!final_script_witness.IsNull()) {
+            SerializeToVector(s, PSBT_IN_SCRIPTWITNESS);
+            SerializeToVector(s, final_script_witness.stack);
+        }
+
+        // Write unknown things
+        for (auto& entry : unknown) {
+            s << entry.first;
+            s << entry.second;
+        }
+
+        s << PSBT_SEPARATOR;
+    }
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        // Read loop
+        bool found_sep = false;
+        while(!s.empty()) {
```

Comment:
Space before `(`. Applies throughout this PR.

Please run all new files through `clang-format` :-)

Reply:
This is moved code, none of this code is new -- I expect I should not actually clang-format it, to preserve move-detection?

Reply:
Agree.

Reply:
Makes sense!